### ISSUE TITLE
[Feature Request]: User Roles - Having Non Admin User

### ISSUE THREAD

**Issue by stratus-ss** [tubearchivist/tubearchivist#529]
### I've read the documentation

- [X] I have read the [how to open an issue](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#how-to-open-an-issue) guide, particularly the [feature request](https://github.com/tubearchivist/tubearchivist/blob/master/CONTRIBUTING.md#feature-request) section.

## Is your feature request related to a problem? Please describe.

It's not a problem exactly but it is on the roadmap.

## Describe the solution you'd like

I would like to have the ability to create a user which is essentially a read only, or consumer only access. This is suitable for children or other access whereby you don't want to allow them to delete or add to the archive

## Additional context

I have some experience in Python (I maintained the Mycroft AI Home Assistant skill for quite some time). I think I could take a good stab at this if someone could take just a bit of time to highlight areas where relevant code might/should/is implemented.

### Your help is needed!

- [X] Yes I will work on this in the next few days or weeks.

**Comment by bbilly1** [comment]
Thanks for looking into this, we've had it on the roadmap for some time, but we were waiting for you. :-)

In any case, I've thought about it a bit before, there are quite a few parts missing for this to become a multi user project:

- Watch history: This currently is not on a per user basis, e.g. we need to have a way to track this in the index by user id. Unclear how to implement that, also current watched history needs to be kept. 
- Personal videos: As in having videos, channels, playlists on a per user basis. So different users should be able to see different videos.
- Permissions: That should be able to be implemented through the django user attributes, superuser and staff. E.g. we could say a user without the staff attribute can only watch, but not edit. That could then be validated in the views and API routes. Then a superuser could be able to add or remove staff permissions of an account in the admin interface. Preferably (I think?) this could be done through extending the base view to check the user object from the request, but a mixin could also achieve the same, and might even be easier to *plug* that in for the views that need to be restricted.

So ideally we'd implement that in the order as outlined above, to make this a real multi user project. But you could also just take on the permission part first, with the obvious limitations from that.

So as you see, that's not yet fully thought through, and take the above with a grain of salt. I'd be curious to hear how you'd implement that! I'm also on discord from time to time.

**Comment by bbilly1** [comment]
take a look at contributing.md, that has some usefull info. there is a GH action for linting in place, you can also run it locally first.

you'll see there is a bit of a hybrid thing going on with some very standard django things, but also a bunch of custom solutions. but the views for permissions should be pretty standard.

and yes, if you get stuck, don't hesitate to reach ou on discord.

**Comment by stratus-ss** [comment]
So I believe we can short cut this fairly easily by just checking if the user is staff. On line 89 of video.html, we can render the template like so:

```
{% if request.user.is_staff %}
          <a download="" href="/media/{{ video.media_url }}"><button id="download-item">Download File</button></a>
                <button onclick="deleteConfirm()" id="delete-item">Delete Video</button>
                <div class="delete-confirm" id="delete-button">
                    <span>Are you sure? </span><button class="danger-button" onclick="deleteVideo(this)" data-id="{{ video.youtube_id }}" data-redirect = "{{ video.channel.channel_id }}">Delete</button> <button onclick="cancelDelete()">C>
                </div>
{% endif %}
```

I haven't hooked up ldap to this or anything. Groups seem like they would be a similar deal. StackOverflow has this example which seems valid
```
from django import template

register = template.Library() 

@register.filter(name='has_group') 
def has_group(user, group_name):
    return user.groups.filter(name=group_name).exists() 
```
In your template:
```
{% if request.user|has_group:"mygroup" %} 
    <p>User belongs to my group 
{% else %}
    <p>User doesn't belong to mygroup</p>
{% endif %}
```

Thoughts?

**Comment by stratus-ss** [comment]
I am proposing that we remove the delete buttons, the add channel buttons and the reindex buttons in order to make a view-only user.

**Comment by stratus-ss** [comment]
submitted https://github.com/tubearchivist/tubearchivist/pull/539

**Comment by bbilly1** [comment]
that's now shipped in v0.4.3. :-)
